MAINT: Hausdorff Generator handling

tylerjereddy opened the pull request:
* This patch is a precursor to SPEC 7 compliance of `directed_hausdorff` requested in gh-21833. I believe the team prefers that I first add support for `Generator` handling here, and once they are happy with that they will follow up with the usual decorator magic.

* In addition to modernization of the source proper to accept a `Generator`, some test cases using a `Generator` with the "same" integer seed as the previous `RandomState` approach have been added. One such case demonstrates that defaulting to `Generator` in the long term will change some results, which SPEC 7 does acknowledge. For clarity, the returned Hausdorff distance will always remain identical, but the indices corresponding to that distance may vary according to the exact random status for this shuffling-reliant algorithm. Disruption is likely to be minimal if we follow the SPEC process/warnings timeline I'd say. Of course, if we follow the full procedure the indices returned for the same function call can change for some inputs.

* Note that there are type stub/`pyi` signature(s) that might in principle be changed here but `mypy` didn't seem to care locally and IIRC our recent philosophy has been to delegate typing improvements to an external package.

[skip cirrus] [skip circle]

tylerjereddy committed:
MAINT: Hausdorff Generator handling

* This patch is a precursor to SPEC 7 compliance of
`directed_hausdorff` requested in gh-21833. I believe
the team prefers that I first add support for `Generator`
handling here, and once they are happy with that they will
follow up with the usual decorator magic.

* In addition to modernization of the source proper to accept a `Generator`,
some test cases using a `Generator` with the "same" integer
seed as the previous `RandomState` approach have been added.
One such case demonstrates that defaulting to `Generator` in the
long term will change some results, which SPEC 7 does acknowledge.
For clarity, the returned Hausdorff distance will always remain identical,
but the indices corresponding to that distance may vary according to
the exact random status for this shuffling-reliant
algorithm. Disruption is likely to be minimal if we follow the SPEC
process/warnings timeline I'd say. Of course, if we follow
the full procedure the indices returned for the same function
call can change for some inputs.

* Note that there are type stub/`pyi` signature(s) that might
in principle be changed here but `mypy` didn't seem to care locally
and IIRC our recent philosophy has been to delegate typing
improvements to an external package.

[skip cirrus] [skip circle]

mdhaber commented on scipy/spatial/_hausdorff.pyx:
```diff
@@ -35,7 +35,11 @@ def directed_hausdorff(const double[:,::1] ar1, const double[:,::1] ar2, seed=0)
     # an advantageous break in the inner search loop and never decreases the
     # performance of the algorithm
-    rng = np.random.RandomState(seed)
+    if isinstance(seed, int | None):
+        rng = np.random.RandomState(seed)
+    else:
+        # Generator passthrough
+        rng = seed
```
There are a few differences between this and the old/usual behavior (e.g. it doesn't look like NumPy integers will be work anymore). To guarantee backward compatibility, I'd suggest branching based on whether `rng` is an instance of `Generator` instead of looking for types that can be passed to `RandomState`.

On the other hand, it would be easiest for the transition if this were to behave like other functions. Can you use [`check_random_state`](https://github.com/scipy/scipy/blob/e24c0cccefb60d927b41ae75a1a772367dd3e340/scipy/_lib/_util.py#L482-L490) or copy that logic here?  It looks like the differences would be:
- `RandomState` (and `Generator`) instances would now be accepted
- If `random_state` is `None` (or `np.random`), it would now be controlled by the global random state `np.random.seed` instead of being unpredictable.

Do you think that is OK?

tylerjereddy replied:
This seems to be getting too confusing IMO.

I intentionally wrote the tests `test_random_state` and `test_random_state_None_int` when this was coded years ago for the reasons noted in the tests there:

> ensure that the global random state is not modified because
> the directed Hausdorff algorithm uses randomization

and

> check that seed values of None or int do not alter global
> random state

I think the original design was meant to keep the function insulated from affecting/consuming external random states.

If I understand Robert correctly, it may make sense to deviate from the SPEC here and keep things mostly as they are with the `seed` argument and just transition that to `Generator`. That said, I'd still prefer to deprecate the switch.

tylerjereddy replied:
(or whatever class of warning is needed... I guess you can't really deprecate `seed=0` changing that easily since it can't be suppressed by user cooperation..

rkern replied:
I would recommend (ultimately) going through with the SPEC 7 conventions (the argument name being `rng`, using a `Generator` internally). I would skip the intermediate stage where we `check_random_state()` can use the legacy global state. If that means that you don't get to use the standard decorator in this one place, that's fine.

mdhaber replied:
I brought this up because the PR breaks backward compatibility immediately.
```python3
seed = np.int64(1)
directed_hausdorff(u, v, seed=seed)  # works before, but not with this PR, failing with
# AttributeError: 'numpy.int64' object has no attribute 'shuffle'
```
To avoid this without looking for all possible types that `seed` used to accept, the first paragraph of my comment is to change this to:
```suggestion
    if isinstance(seed, np.random.Generator):
    # Generator passthrough
        rng = seed
    else:
        rng = np.random.RandomState(seed)
```

---

> I think the original design was meant to keep the function insulated from affecting/consuming external random states.

If that is a very important property to preserve, that's fine. If it were *not* important to preserve that, I was suggesting that this be temporarily simplified (during the SPEC 7 deprecation period) to:
```suggestion
    rng = check_random_state(seed)
```
for the reason that it would make the SPEC 7 transition simpler.

But again, I don't know anything about this function or the existing tests, so if it's important to not affect/consume the external random state, that's fine. I just wanted to explore options to keep things easier for us.

rkern replied:
Yes, it's an important property to preserve.

mdhaber commented on scipy/spatial/tests/test_hausdorff.py:
```diff
@@ -133,6 +133,15 @@ def test_invalid_dimensions(self):
+        ([(0,0)],
+         [(0,1), (0,0)],
+         np.random.default_rng(1),
+         (0.0, 0, 1)),
```
Where do the reference values come from?

tylerjereddy replied:
Two things:

1) It is a direct copy paste of an already-tested case above, with the adjustment of using a `Generator` instead of an integer seed.
2) The expected value can be deduced by straightforward reasoning--the "maximum minimum" distance between `u` and `v` must be 0 since there's an identical point in the two "curves."

tylerjereddy replied:
And note that this is asymmetric, in the reverse order (`v` to `u`) there are two points in `v` and the "maximum minimum" becomes `1`.

mdhaber reviewed (commented):
Thanks @tylerjereddy! This looks pretty straightforward; just two questions.

rkern commented:
> * Of course, if we follow the full procedure the indices returned for the same function call can change for some inputs.

FWIW, long term, I think it's fine for this function to be an exception to the rule for the default value. This is one of those uses of pseudorandomness where we're not looking for its statistical properties so much as (almost certainly) avoiding worst-case algorithmic performance. The current semantics (leaving the `seed` as the default means that the same input gives the same output within the same program run) are desirable for this use case.

I do think it should use `Generator` and accept that different runs of the program might yield different results from version to version of numpy and scipy, despite the same default integer seed. That amount of stability isn't worth remaining on `RandomState`, IMO, for all the reasons laid out in NEP 19.

mdhaber reviewed (approved):
Thanks @tylerjereddy. This looks fine with me with the one modification - the first suggestion in https://github.com/scipy/scipy/pull/21897/files#r1845257793. If that looks good to you, please commit it, and I can merge when tests pass.

tylerjereddy committed:
BUG: PR 21897 revisions

* Based on reviewer feedback, add a regression test
that enforces 0-D NumPy integer array support
for `directed_hausdorff` `seed` argument, since this
was previously supported.

* Apply a source code fix for the above issue.

[skip cirrus]

tylerjereddy commented:
@mdhaber Ok, I think I've revised accordingly now, including a regression test for supporting a 0-D NumPy integer array provided to `seed`.

If a SPEC-related PR happens for `directed_hausdorff` after this please do ping me.

mdhaber commented:
Yup, I'll ping you. LGTM and I'll merge when tests pass. Thanks!
